nemo_rl/distributed/named_sharding.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, List, Sequence, Union
+
+import numpy as np
+
+
+class NamedSharding:
+    """Represents an N-dimensional arrangement of ranks with named axes, facilitating data sharding, replication, and collection based on these axes.
+
+    Example:
+        layout = [
+            [[0, 1, 2, 3], [4, 5, 6, 7]],
+        ]
+        names = ["dp", "pp", "tp"]
+        # This represents DP=1, PP=2, TP=4
+        sharding = NamedSharding(layout, names)
+        print(sharding.shape)  # Output: (1, 2, 4)
+        print(sharding.names)  # Output: ['dp', 'pp', 'tp']
+        print(sharding.get_ranks(dp=0, pp=1)) # Output: [4, 5, 6, 7]
+    """
+
+    def __init__(self, layout: Sequence[Any], names: List[str]):
```

Comment:
wdyt about the API being the sharding itself?
```suggestion
    def __init__(self, **name_to_sharding_dim):
```

the user needs to make sure the sharding is correct and we have to do a bunch of checks, but if we only let them specify the sizes, it seems like we can make the logic simpler and guarantee the ordering.

Reply:
I considered that, but didn't like how it made it less obvious how the ranks were arranged along axes. The ordering of these axes is very important and getting the groups is trivial when you have the full layout as a tensor. if you do it as a set of axes and sizes, then you have to start to do arithmetic to figure out how to work with it.

The upside though is that you don't have this tensor anymore and just have some sizes.
 Both are acceptable IMO.

Reply:
I prefer the explicit api that's currently implemented (for the same reasons @SahilJain314 mentioned about how ranks are arranged)

Reply:
So currently the NamedSharding instantiation looks like this:

```python
        self.sharding_annotations = NamedSharding(
            layout=np.arange(cluster.world_size()).reshape(
                -1,  # DP
                config["vllm_cfg"]["tensor_parallel_size"],  # TP
            ),
            names=["data_parallel", "tensor_parallel"],
        )
```
are you saying that it's easier to see that that tensor layout is being passed in? Also, I don't understand your comment about arithmetic here; could you elaborate?

Reply:
yeah, the tensor layout is easier to see (with explicit coordinate -> rank mapping). On arithmetic, you'd need to do some basic stride arithmetic to figure out which ranks exist for a set of axes or the other way around. i.e. finding coords would require some arithmetic where you calculate cumulative strides from the axis sizes and work from that.